# Tri-State Glory

We're tasked with choosing the best zip codes in the tri-state area to buy in. We made the decision by looking at the return on investement between 1996 and 2018. From there we landed on our 9 zip codes that we'll further explore in subsequent notebooks. This notebook attains the original Zillow data and filters down to the zip codes we want to investigate.

The first step is to import the necessary libraries to get the data.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import missingno as mno
import warnings
warnings.filterwarnings('ignore')

In [3]:
pd.set_option('display.max_columns', None)
# setting format for readability
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [4]:
# read in the csv

df = pd.read_csv('zillow_data.csv')
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08   1996-09  \
0         1 334200.00 335400.00 336500.00 337600.00 338500.00 339500.00   
1         2 235700.00 236900.00 236700.00 235400.00 233300.00 230600.00   
2         3 210400.00 212200.00 212200.00 210700.00 208300.00 205500.00   
3         4 498100.00 500900.00 503100.00 504600.00 505500.00 505700.00   
4         5  77300.00  77300.00  77300.00  77300.00  77400.00  77500.00   

    1996-10   1996-11   1996-12   1997-01   1997-02   1997-03   1997-04  \
0 340400.00 341300.00 342600.00 344400.00 345700.00 346700.00 347800.00   
1 227300.00 223400.00 219600.00 215800.00 211100.00 205700.00 200900.00   
2 202500.00 199800.00 198300.00 197300.00 195400.00 193000.00 191800.00   
3 505300.00 504200.00 503600.00 503400.00 502200.00 500000.00 497900.00   
4  77600.00  77700.00  77700.00  77800.00  77900.00  77900.00  77800.00   

    1997-05   1997-06   1997-07   1997-08   1997-09   1997-10   1997-11  \
0 349000.00 350400.00 352000.00 353900.00 356200.00 358800.00 361800.00   
1 196800.00 193600.00 191400.00 190400.00 190800.00 192700.00 196000.00   
2 191800.00 193000.00 195200.00 198400.00 202800.00 208000.00 213800.00   
3 496300.00 495200.00 494700.00 494900.00 496200.00 498600.00 502000.00   
4  77800.00  77800.00  77800.00  77800.00  77900.00  78100.00  78200.00   

    1997-12   1998-01   1998-02   1998-03   1998-04   1998-05   1998-06  \
0 365700.00 370200.00 374700.00 378900.00 383500.00 388300.00 393300.00   
1 201300.00 207400.00 212200.00 214600.00 215100.00 213400.00 210200.00   
2 220700.00 227500.00 231800.00 233400.00 233900.00 233500.00 233300.00   
3 507600.00 514900.00 522200.00 529500.00 537900.00 546900.00 556400.00   
4  78400.00  78600.00  78800.00  79000.00  79100.00  79200.00  79300.00   

    1998-07   1998-08   1998-09   1998-10   1998-11   1998-12   1999-01  \
0 398500.00 403800.00 409100.00 414600.00 420100.00 426200.00 432600.00   
1 206100.00 202100.00 198800.00 196100.00 194100.00 193400.00 193400.00   
2 234300.00 237400.00 242800.00 250200.00 258600.00 268000.00 277000.00   
3 566100.00 575600.00 584800.00 593500.00 601600.00 610100.00 618600.00   
4  79300.00  79300.00  79400.00  79500.00  79500.00  79600.00  79700.00   

    1999-02   1999-03   1999-04   1999-05   1999-06   1999-07   1999-08  \
0 438600.00 444200.00 450000.00 455900.00 462100.00 468500.00 475300.00   
1 193100.00 192700.00 193000.00 193700.00 194800.00 196100.00 197800.00   
2 283600.00 288500.00 293900.00 299200.00 304300.00 308600.00 311400.00   
3 625600.00 631100.00 636600.00 642100.00 647600.00 653300.00 659300.00   
4  79900.00  80100.00  80300.00  80600.00  80900.00  81200.00  81400.00   

    1999-09   1999-10   1999-11   1999-12   2000-01   2000-02   2000-03  \
0 482500.00 490200.00 498200.00 507200.00 516800.00 526300.00 535300.00   
1 199700.00 201900.00 204500.00 207800.00 211500.00 214900.00 217800.00   
2 312300.00 311900.00 311100.00 311700.00 313500.00 315000.00 316700.00   
3 665800.00 672900.00 680500.00 689600.00 699700.00 709300.00 718300.00   
4  81700.00  82100.00  82400.00  82600.00  82800.00  82900.00  83000.00   

    2000-04   2000-05   2000-06   2000-07   2000-08   2000-09   2000-10  \
0 544500.00 553500.00 562400.00 571200.00 579800.00 588100.00 596300.00   
1 221100.00 224100.00 226700.00 228200.00 228500.00 227200.00 224900.00   
2 319800.00 323700.00 327500.00 329900.00 329800.00 326400.00 320100.00   
3 727600.00 737100.00 746600.00 756200.00 765800.00 775100.00 784400.00   
4  83000.00  82900.00  82800.00 

## Data Preprocessing

The Zillow file reads in as wide format. We'll need to melt it down to long format to do our time series analysis, but that will happen in a separate notebook. The first thing we want to do is change the datatype for all of the observations starting in April 1996. This function has been provided to us by the Flatiron School.

In [5]:
# provided function

def get_datetimes(df):
    return pd.to_datetime(df.columns.values[7:], format='%Y-%m')

In [6]:
# pass Zillow data through get_datetimes function

get_datetimes(df)

DatetimeIndex(['1996-04-01', '1996-05-01', '1996-06-01', '1996-07-01',
               '1996-08-01', '1996-09-01', '1996-10-01', '1996-11-01',
               '1996-12-01', '1997-01-01',
               ...
               '2017-07-01', '2017-08-01', '2017-09-01', '2017-10-01',
               '2017-11-01', '2017-12-01', '2018-01-01', '2018-02-01',
               '2018-03-01', '2018-04-01'],
              dtype='datetime64[ns]', length=265, freq=None)

In [7]:
# checking the shape of our data

df.shape

(14723, 272)

We have 272 observed zip codes from 1996 and 2018 across the United States. We'll be focusing on the tristate area which we filter for below.

In order to determine the 'best' zip codes, we want to see where the best return is. We added the ROI column to determine which zip codes we want to model for.

In [8]:
df['ROI'] = (df['2018-04']/df['1996-04'])-1

## Filtering for Zipcodes

We're interesting at looking only in the tristate area. We filter our dataframe for just NY, NJ and CT. Once we filter, we see that we've cut our number of zip codes observed from 14.7k to 1.6k. We also want to check for null values.

In [9]:
tristate = df[df['State'].isin(['NY', 'NJ', 'CT'])]
tristate.shape

(1641, 273)

We have null values in the Metro column and the ROI column. We're not concerned with the Metro column in this project, but we do need to understand the null values in the ROI column. When we filtered for those values we saw that there is no recored data for these 30 zip codes from 1996 to 2004. We made the decision to only look at values that have had observations for the full time period and dropped these observations.

In [12]:
tristate.isnull().sum()

RegionID       0
RegionName     0
City           0
State          0
Metro         52
              ..
2018-01        0
2018-02        0
2018-03        0
2018-04        0
ROI           30
Length: 273, dtype: int64

In [13]:
tristate[tristate['ROI'].isnull()]

RegionID  RegionName          City State       Metro   CountyName  \
20        61625       10011      New York    NY    New York     New York   
36        61796       10456      New York    NY    New York        Bronx   
156       62048       11238      New York    NY    New York        Kings   
272       61635       10021      New York    NY    New York     New York   
508       61628       10014      New York    NY    New York     New York   
1299      62004       11101      New York    NY    New York       Queens   
1413      61799       10459      New York    NY    New York        Bronx   
1434      62016       11205      New York    NY    New York        Kings   
1534      62028       11217      New York    NY    New York        Kings   
1995      62041       11231      New York    NY    New York     New York   
2729      61857       10550  Mount Vernon    NY    New York  Westchester   
6640      63479       14222       Buffalo    NY     Buffalo         Erie   
7304      63458       14201       Buffalo    NY     Buffalo         Erie   
9396      63372       14032      Clarence    NY     Buffalo         Erie   
9591      62555       12428     Wawarsing    NY    Kingston       Ulster   
9942      62292       11944     Greenport    NY    New York      Suffolk   
10136     62753       12790    Mamakating    NY         NaN     Sullivan   
11293     63406       14080       Holland    NY     Buffalo         Erie   
12137     62738       12775     Rock Hill    NY         NaN     Sullivan   
12546     63171       13607    Alexandria    NY   Watertown    Jefferson   
12565     62419       12138    Petersburg    NY      Albany   Rensselaer   
12588     63365       14025        Boston    NY     Buffalo         Erie   
12726     62582       12458     Wawarsing    NY    Kingston       Ulster   
13176     62225       11749         Islip    NY    New York      Suffolk   
13222     63215       13656       Orleans    NY   Watertown    Jefferson   
13369     63373       14033        Colden    NY     Buffalo         Erie   
13439     63678       14716       Brocton    NY   Jamestown   Chautauqua   
13790     63300       13797         Lisle    NY  Binghamton       Broome   
13931     62746       12783        Bethel    NY         NaN     Sullivan   
14327     63370       14030      Sardinia    NY     Buffalo         Erie   

       SizeRank  1996-04  1996-05  1996-06  1996-07  1996-08  1996-09  \
20           21      nan      nan      nan      nan      nan      nan   
36           37      nan      nan      nan      nan      nan      nan   
156         157      nan      nan      nan      nan      nan      nan   
272         273      nan      nan      nan      nan      nan      nan   
508         509      nan      nan      nan      nan      nan      nan   
1299       1300      nan      nan      nan      nan      nan      nan   
1413       1414      nan      nan      nan      nan      nan      nan   
1434       1435      nan      nan      nan      nan      nan      nan   
1534       1535      nan      nan      nan      nan      nan      nan   
1995       1996      nan      nan      nan      nan      nan      nan   
2729       2730      nan      nan      nan      nan      nan      nan   
6640       6641      nan      nan      nan      nan      nan      nan   
7304       7305      nan      nan      nan      nan      nan      nan   
9396       9397      nan      nan      nan      nan      nan      nan   
9591       9592      nan      nan      nan      nan      nan      nan   
9942       9943      nan      nan      nan      nan      nan      nan   
10136     10137      nan      nan      nan      nan      nan      nan   
11293     11294      nan      nan      nan      nan      nan      nan   
12137     12138      nan      nan      nan      nan      nan      nan   
12546     12547      nan      nan      nan      nan      nan      nan   
12565     12566      nan      nan      nan      nan      nan      nan   
12588     12589      nan      nan      nan  

In [20]:
tristate.dropna(subset=['ROI'],inplace=True)

To keep as a reference, we added the 'AveragePrice' column to see the average price over all the time period. This is just for reference to gain more domain knowledge on the data we're working with. This column will dropped once we melt the data into the long form and begin modeling.

In [21]:
tristate['AveragePrice'] = tristate.loc[:,'1996-04':'2018-04'].mean(axis=1)

In [22]:
tristate.shape

(1611, 274)

In [23]:
# rename RegionName to Zipcode to be more interpretable

tristate.rename({'RegionName': 'Zipcode'}, axis='columns', inplace=True)

We first wanted to check the top 5 zip codes based on the ROI. When we filtered for this, it looks very similiar - lots of Brooklyn only interrupted by Jersey City. We want to offer a more robust look at the tristate area. Instead of only looking at the top 5 zip codes, we decided to look at the top 3 in each state.

In [24]:
tristate.sort_values('ROI',axis=0,ascending=False)[:5]

RegionID  Zipcode         City State     Metro CountyName  SizeRank  \
117      62022    11211     New York    NY  New York      Kings       118   
1155     62033    11222     New York    NY  New York      Kings      1156   
475      62027    11216     New York    NY  New York      Kings       476   
191      60639     7302  Jersey City    NJ  New York     Hudson       192   
106      62026    11215     New York    NY  New York      Kings       107   

       1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
117  133200.00 132900.00 132500.00 132200.00 131800.00 131600.00 131500.00   
1155 149200.00 148400.00 147500.00 146600.00 146000.00 145600.00 145400.00   
475  146100.00 146600.00 147200.00 147700.00 148400.00 149300.00 150200.00   
191  137200.00 137800.00 138500.00 139100.00 139600.00 140100.00 140700.00   
106  225700.00 227500.00 229400.00 231600.00 234100.00 236900.00 239800.00   

       1996-11   1996-12   1997-01   1997-02   1997-03   1997-04   1997-05  \
117  131800.00 132400.00 133100.00 133600.00 134100.00 134400.00 134700.00   
1155 145600.00 146100.00 146600.00 146900.00 147100.00 147200.00 147200.00   
475  151300.00 152700.00 154100.00 155200.00 156200.00 157000.00 157800.00   
191  141400.00 142300.00 143500.00 144600.00 145500.00 146400.00 147400.00   
106  242900.00 246400.00 249800.00 252200.00 253800.00 255100.00 256000.00   

       1997-06   1997-07   1997-08   1997-09   1997-10   1997-11   1997-12  \
117  135000.00 135400.00 135900.00 136400.00 137200.00 138300.00 139800.00   
1155 147200.00 147400.00 147800.00 148500.00 149400.00 150800.00 152400.00   
475  158400.00 159100.00 159800.00 160700.00 161700.00 162800.00 164000.00   
191  148300.00 149100.00 150000.00 150900.00 151800.00 152800.00 154300.00   
106  256900.00 258200.00 260000.00 262300.00 264900.00 268000.00 271900.00   

       1998-01   1998-02   1998-03   1998-04   1998-05   1998-06   1998-07  \
117  141500.00 142900.00 144200.00 145300.00 146200.00 147000.00 147700.00   
1155 154100.00 155500.00 156600.00 157500.00 158200.00 158800.00 159500.00   
475  165300.00 166300.00 167100.00 167800.00 168400.00 168900.00 169500.00   
191  156100.00 157900.00 159400.00 161100.00 162700.00 164500.00 166200.00   
106  276000.00 279300.00 282000.00 284700.00 287300.00 290100.00 293700.00   

       1998-08   1998-09   1998-10   1998-11   1998-12   1999-01   1999-02  \
117  148500.00 149300.00 150200.00 151400.00 153100.00 154800.00 156300.00   
1155 160400.00 161500.00 162900.00 164800.00 167100.00 169300.00 171300.00   
475  170200.00 171200.00 172500.00 173900.00 175600.00 177500.00 179200.00   
191  167900.00 169700.00 171600.00 173600.00 176000.00 178800.00 181600.00   
106  297900.00 302800.00 308200.00 314100.00 321000.00 327800.00 333700.00   

       1999-03   1999-04   1999-05   1999-06   1999-07   1999-08   1999-09  \
117  157500.00 158600.00 159500.00 160300.00 161100.00 162100.00 163100.00   
1155 173000.00 174700.00 176200.00 177700.00 179400.00 181500.00 184000.00   
475  180800.00 182300.00 183800.00 185400.00 187000.00 188900.00 191000.00   
191  184100.00 186800.00 189500.00 192300.00 195000.00 197700.00 200400.00   
106  338700.00 343500.00 348100.00 352700.00 358100.00 364400.00 371400.00   

       1999-10   1999-11   1999-12   2000-01   2000-02   2000-03   2000-04  \
117  164400.00 166100.00 168400.00 170900.00 173300.00 175400.00 177500.00   
1155 186800.00 190300.00 194200.00 198200.00 201700.00 204900.00 207800.00   
475  193400.00 196000.00 198900.00 201900.00 204700.00 207300.00 209600.00   
191  203200.00 206100.00 209400.00 213300.00 217200.00 220900.00 224700.00   
106  378800.00 386900.00 396000.00 404900.00 412600.00 419200.00 425200.00   

       2000-05   2000-06   2000-07   2000-08   2000-09   2000-10   2000-11  \
117  179400.00 181200.00 183100.00 185200.00 187500.00 190000.00 193100.00   
1155 210300.00 212500.00 214900.00 217500.00 220500.00 223700.00 227400.00   
475  211800.00 213800.00 2

In [25]:
# saving each state as its own dataframe to then filter for

new_york = tristate[tristate['State'] == 'NY']
new_jersey = tristate[tristate['State'] == 'NJ']
connecticut = tristate[tristate['State'] == 'CT']

In [26]:
new_york[['Zipcode', 'City', 'State', 'Metro', 'CountyName', 'AveragePrice', 'ROI']].sort_values('ROI',axis=0,ascending=False)[:3]

Zipcode      City State     Metro CountyName  AveragePrice   ROI
117     11211  New York    NY  New York      Kings     595167.17 11.19
1155    11222  New York    NY  New York      Kings     640047.55 10.54
475     11216  New York    NY  New York      Kings     572171.70  9.94

The top 3 zip codes for New York are all in Brooklyn. 11211 is Williamsburg. 11222 is Greenpoint and 11216 covers part of Bedford-Stuyvesant next to Clinton Hill and Stuyvesant Heights.

In [22]:
new_jersey[['Zipcode', 'City', 'State', 'Metro', 'CountyName', 'AveragePrice', 'ROI']].sort_values('ROI',axis=0,ascending=False)[:3]

Zipcode         City State     Metro CountyName  AveragePrice  ROI
191      7302  Jersey City    NJ  New York     Hudson     596591.32 9.40
57       7030      Hoboken    NJ  New York     Hudson     957777.36 5.15
9758     7756      Neptune    NJ  New York   Monmouth     345644.53 4.42

For New Jesery, 07302 Jersey City has the highest ROI, followed by 07030 Hoboken. Behind Hoboken is the Jesery Shore town 07756 Ocean Grove.

In [23]:
connecticut[['Zipcode', 'City', 'State', 'Metro', 'CountyName', 'AveragePrice', 'ROI']].sort_values('ROI',axis=0,ascending=False)[:3]

Zipcode        City State       Metro  CountyName  AveragePrice  ROI
6036      6820      Darien    CT    Stamford   Fairfield    1128919.62 1.51
13645     6068   Salisbury    CT  Torrington  Litchfield     384341.13 1.48
14355     6793  Washington    CT  Torrington  Litchfield     583974.72 1.37

Unlike New York and New Jersey, the zip codes in Connecticut that had the biggest ROI were not cities. The ROI was much more modest in Connecticut compared to the other two states. 06820 is Darien, CT is located right near Stamford along the Long Island Sound. 06068 is Salisbury, CT located in the corner of CT along the Massachusetts and New York border. The final zip code for Connecticut is 06793, Washington, CT also near the New York border.

In [27]:
# save new dataframe for the 9 zip codes we want to model for later

zip_codes = [11211, 11222, 11216, 7302, 7030, 7756, 6820, 6068, 6793]

zipcodes = tristate[tristate['Zipcode'].isin(zip_codes)]
zipcodes.head()

RegionID  Zipcode         City State     Metro CountyName  SizeRank  \
57       60545     7030      Hoboken    NJ  New York     Hudson        58   
117      62022    11211     New York    NY  New York      Kings       118   
191      60639     7302  Jersey City    NJ  New York     Hudson       192   
475      62027    11216     New York    NY  New York      Kings       476   
1155     62033    11222     New York    NY  New York      Kings      1156   

       1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
57   301600.00 302700.00 303800.00 304800.00 305800.00 306800.00 308100.00   
117  133200.00 132900.00 132500.00 132200.00 131800.00 131600.00 131500.00   
191  137200.00 137800.00 138500.00 139100.00 139600.00 140100.00 140700.00   
475  146100.00 146600.00 147200.00 147700.00 148400.00 149300.00 150200.00   
1155 149200.00 148400.00 147500.00 146600.00 146000.00 145600.00 145400.00   

       1996-11   1996-12   1997-01   1997-02   1997-03   1997-04   1997-05  \
57   309400.00 311400.00 314000.00 316400.00 318400.00 320600.00 322900.00   
117  131800.00 132400.00 133100.00 133600.00 134100.00 134400.00 134700.00   
191  141400.00 142300.00 143500.00 144600.00 145500.00 146400.00 147400.00   
475  151300.00 152700.00 154100.00 155200.00 156200.00 157000.00 157800.00   
1155 145600.00 146100.00 146600.00 146900.00 147100.00 147200.00 147200.00   

       1997-06   1997-07   1997-08   1997-09   1997-10   1997-11   1997-12  \
57   325300.00 327700.00 330300.00 332900.00 335800.00 339000.00 342900.00   
117  135000.00 135400.00 135900.00 136400.00 137200.00 138300.00 139800.00   
191  148300.00 149100.00 150000.00 150900.00 151800.00 152800.00 154300.00   
475  158400.00 159100.00 159800.00 160700.00 161700.00 162800.00 164000.00   
1155 147200.00 147400.00 147800.00 148500.00 149400.00 150800.00 152400.00   

       1998-01   1998-02   1998-03   1998-04   1998-05   1998-06   1998-07  \
57   347400.00 351800.00 355900.00 360200.00 364700.00 369200.00 373800.00   
117  141500.00 142900.00 144200.00 145300.00 146200.00 147000.00 147700.00   
191  156100.00 157900.00 159400.00 161100.00 162700.00 164500.00 166200.00   
475  165300.00 166300.00 167100.00 167800.00 168400.00 168900.00 169500.00   
1155 154100.00 155500.00 156600.00 157500.00 158200.00 158800.00 159500.00   

       1998-08   1998-09   1998-10   1998-11   1998-12   1999-01   1999-02  \
57   378400.00 383100.00 388100.00 393300.00 399300.00 405600.00 411800.00   
117  148500.00 149300.00 150200.00 151400.00 153100.00 154800.00 156300.00   
191  167900.00 169700.00 171600.00 173600.00 176000.00 178800.00 181600.00   
475  170200.00 171200.00 172500.00 173900.00 175600.00 177500.00 179200.00   
1155 160400.00 161500.00 162900.00 164800.00 167100.00 169300.00 171300.00   

       1999-03   1999-04   1999-05   1999-06   1999-07   1999-08   1999-09  \
57   417600.00 423600.00 429400.00 435000.00 440500.00 445800.00 451000.00   
117  157500.00 158600.00 159500.00 160300.00 161100.00 162100.00 163100.00   
191  184100.00 186800.00 189500.00 192300.00 195000.00 197700.00 200400.00   
475  180800.00 182300.00 183800.00 185400.00 187000.00 188900.00 191000.00   
1155 173000.00 174700.00 176200.00 177700.00 179400.00 181500.00 184000.00   

       1999-10   1999-11   1999-12   2000-01   2000-02   2000-03   2000-04  \
57   456500.00 462100.00 468500.00 475300.00 482100.00 488500.00 495000.00   
117  164400.00 166100.00 168400.00 170900.00 173300.00 175400.00 177500.00   
191  203200.00 206100.00 209400.00 213300.00 217200.00 220900.00 224700.00   
475  193400.00 196000.00 198900.00 201900.00 204700.00 207300.00 209600.00   
1155 186800.00 190300.00 194200.00 198200.00 201700.00 204900.00 207800.00   

       2000-05   2000-06   2000-07   2000-08   2000-09   2000-10   2000-11  \
57   501400.00 507800.00 514200.00 520400.00 526800.00 533400.00 540400.00   
117  179400.00 181200.00 183100.00 185200.00 187500.00 190000.00 193100.00   
191  228700.00 232700.00 2

The final step is to save the data as a new csv. We'll work off of this csv to melt our data and later model for furture values for each individual zip codes.

In [28]:
zipcodes.to_csv('zipcode_data.csv')